File: app/src/main/java/com/google/samples/apps/topeka/adapter/ScoreAdapter.java
/*
 * Copyright 2015 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.samples.apps.topeka.adapter;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.annotation.ColorRes;
import android.support.annotation.DrawableRes;
import android.support.v4.content.ContextCompat;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.samples.apps.topeka.R;
import com.google.samples.apps.topeka.model.Category;
import com.google.samples.apps.topeka.model.quiz.Quiz;

import java.util.List;

/**
 * Adapter for displaying score cards.
 */
public class ScoreAdapter extends BaseAdapter {

    private final Category mCategory;
    private final int count;
    private final List<Quiz> mQuizList;

    private Drawable mSuccessIcon;
    private Drawable mFailedIcon;

    public ScoreAdapter(Category category) {
        mCategory = category;
        mQuizList = mCategory.getQuizzes();
        count = mQuizList.size();
    }

    @Override
    public int getCount() {
        return count;
    }

    @Override
    public Quiz getItem(int position) {
        return mQuizList.get(position);
    }

    @Override
    public long getItemId(int position) {
        if (position > count || position < 0) {
            return AbsListView.INVALID_POSITION;
        }
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        if (null == convertView) {
            convertView = createView(parent);
        }

        final Quiz quiz = getItem(position);
        ViewHolder viewHolder = (ViewHolder) convertView.getTag();
        viewHolder.mQuizView.setText(quiz.getQuestion());
        viewHolder.mAnswerView.setText(quiz.getStringAnswer());
        setSolvedStateForQuiz(viewHolder.mSolvedState, position);
        return convertView;
    }

    private void setSolvedStateForQuiz(ImageView solvedState, int position) {
        final Context context = solvedState.getContext();
        final Drawable tintedImage;
        if (mCategory.isSolvedCorrectly(getItem(position))) {
            tintedImage = getSuccessIcon(context);
        } else {
            tintedImage = getFailedIcon(context);
        }
        solvedState.setImageDrawable(tintedImage);
    }

    private Drawable getSuccessIcon(Context context) {
        if (null == mSuccessIcon) {
            mSuccessIcon = loadAndTint(context, R.drawable.ic_tick, R.color.theme_green_primary);
        }
        return mSuccessIcon;
    }

    private Drawable getFailedIcon(Context context) {
        if (null == mFailedIcon) {
            mFailedIcon = loadAndTint(context, R.drawable.ic_cross, R.color.theme_red_primary);
        }
        return mFailedIcon;
    }

    /**
     * Convenience method to aid tintint of vector drawables at runtime.
     *
     * @param context The {@link Context} for this app.
     * @param drawableId The id of the drawable to load.
     * @param tintColor The tint to apply.
     * @return The tinted drawable.
     */
    private Drawable loadAndTint(Context context, @DrawableRes int drawableId,
                                 @ColorRes int tintColor) {
        Drawable imageDrawable = ContextCompat.getDrawable(context, drawableId);
        if (imageDrawable == null) {
            throw new IllegalArgumentException("The drawable with id " + drawableId
                    + " does not exist");
        }
        DrawableCompat.setTint(DrawableCompat.wrap(imageDrawable), tintColor);
        return imageDrawable;
    }

    private View createView(ViewGroup parent) {
        View convertView;
        final LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        ViewGroup scorecardItem = (ViewGroup) inflater.inflate(
                R.layout.item_scorecard, parent, false);
        convertView = scorecardItem;
        ViewHolder holder = new ViewHolder(scorecardItem);
        convertView.setTag(holder);
        return convertView;
    }

    private class ViewHolder {

        final TextView mAnswerView;
        final TextView mQuizView;
        final ImageView mSolvedState;

        public ViewHolder(ViewGroup scorecardItem) {
            mQuizView = (TextView) scorecardItem.findViewById(R.id.quiz);
            mAnswerView = (TextView) scorecardItem.findViewById(R.id.answer);
            mSolvedState = (ImageView) scorecardItem.findViewById(R.id.solved_state);
        }

    }
}

File: quiz/src/main/java/com/google/samples/apps/topeka/adapter/ScoreAdapter.kt
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.samples.apps.topeka.adapter

import android.content.Context
import android.graphics.drawable.Drawable
import android.support.annotation.ColorRes
import android.support.annotation.DrawableRes
import android.support.v4.content.ContextCompat
import android.support.v4.graphics.drawable.DrawableCompat
import android.view.View
import android.view.ViewGroup
import android.widget.AbsListView
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.TextView
import com.google.samples.apps.topeka.quiz.R
import com.google.samples.apps.topeka.helper.inflate
import com.google.samples.apps.topeka.model.Category

/**
 * Adapter for displaying score cards.
 */
class ScoreAdapter(private val context: Context, private val category: Category) : BaseAdapter() {

    private val quizList = category.quizzes
    private val count = quizList.size
    private val successIcon by lazy {
        loadAndTint(
            com.google.samples.apps.topeka.base.R.drawable.ic_tick,
            com.google.samples.apps.topeka.base.R.color.theme_green_primary)
    }
    private val failedIcon by lazy {
        loadAndTint(com.google.samples.apps.topeka.base.R.drawable.ic_cross,
                com.google.samples.apps.topeka.base.R.color.theme_red_primary)
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup) =
            (convertView ?: createView(parent))
                    .apply {
                        with(tag as ViewHolder) {
                            with(getItem(position)) {
                                quizView.text = question
                                answerView.text = stringAnswer
                            }
                            setSolvedStateForQuiz(solvedState, position)
                        }
                    }

    private fun setSolvedStateForQuiz(solvedState: ImageView, position: Int) {
        val tintedImage = if (category.isSolvedCorrectly(getItem(position))) successIcon
        else failedIcon
        solvedState.setImageDrawable(tintedImage)
    }

    override fun getCount() = count

    override fun getItem(position: Int) = quizList[position]

    override fun getItemId(position: Int): Long {
        val itemId = if (position in 0..count + 1) position else AbsListView.INVALID_POSITION
        return itemId.toLong()
    }

    /**
     * Convenience method to aid tint of vector drawables at runtime.

     * @param drawableId The id of the drawable to load.
     *
     * @param tintColor The tint to apply.
     *
     * @return The tinted drawable.
     */
    private fun loadAndTint(@DrawableRes drawableId: Int, @ColorRes tintColor: Int): Drawable {
        val drawable = ContextCompat.getDrawable(context, drawableId)!!
        return DrawableCompat.wrap(drawable)
                .apply { DrawableCompat.setTint(this, ContextCompat.getColor(context, tintColor)) }
    }

    private fun createView(parent: ViewGroup) =
            (parent.context.inflate(R.layout.item_scorecard, parent, false) as ViewGroup)
                    .apply { tag = ViewHolder(this) }

    private inner class ViewHolder(scorecardItem: ViewGroup) {

        internal val answerView = scorecardItem.findViewById<TextView>(R.id.answer)
        internal val quizView = scorecardItem.findViewById<TextView>(R.id.quiz)
        internal val solvedState = scorecardItem.findViewById<ImageView>(R.id.solved_state)

    }
}
